src/graph.rs

```diff
@@ -0,0 +1,221 @@
+        && unit_props1.main_chain_index != None
+    {
+        return Ok(Some(-1));
+    }
+
+    if unit_props1.level <= unit_props2.level
+        && unit_props1.latest_included_mc_index <= unit_props2.latest_included_mc_index
+        && (unit_props1.main_chain_index <= unit_props2.main_chain_index
+            && unit_props1.main_chain_index != None
+            && unit_props2.main_chain_index != None
+            || unit_props1.main_chain_index == None
+            || unit_props2.main_chain_index == None)
+        || unit_props1.level >= unit_props2.level
+            && unit_props1.latest_included_mc_index >= unit_props2.latest_included_mc_index
+            && (unit_props1.main_chain_index >= unit_props2.main_chain_index
+                && unit_props1.main_chain_index != None
+                && unit_props2.main_chain_index != None
+                || unit_props1.main_chain_index == None
+                || unit_props2.main_chain_index == None)
+    {
+        let (earlier_unit, later_unit, result_if_found) = if unit_props1.level < unit_props2.level {
+            (unit_props1, unit_props2, -1)
+        } else {
+            (unit_props2, unit_props1, 1)
+        };
+
+        // In JS it is okay if main_chain_index === null, the delta will be negative, need to check
+        // the None value in rust
+        let earlier_unit_delta =
+            earlier_unit.main_chain_index.unwrap() - earlier_unit.latest_included_mc_index.unwrap();
+        let later_unit_delta =
+            later_unit.main_chain_index.unwrap() - later_unit.latest_included_mc_index.unwrap();
+
+        let mut start_units = Vec::new();
+        if later_unit_delta > earlier_unit_delta {
+            //GoUp()
+            start_units.push(later_unit.unit.clone());
+
+            'go_up: loop {
+                let start_unit_list = start_units
+                    .iter()
+                    .map(|s| format!("'{}'", s))
+                    .collect::<Vec<_>>()
+                    .join(", ");
+
+                let sql = format!(
+                    "SELECT unit, level, latest_included_mc_index, main_chain_index, is_on_main_chain \
+                    FROM parenthoods JOIN units ON parent_unit=unit \
+                    WHERE child_unit IN({})",
+                    start_unit_list
+                );
+
+                let mut stmt = db.prepare_cached(&sql)?;
+                let rows = stmt.query_map(&[], |row| UnitProps {
+                    unit: row.get(0),
+                    level: row.get(1),
+                    latest_included_mc_index: row.get(2),
+                    main_chain_index: row.get(3),
+                    is_on_main_chain: row.get(4),
+                    is_free: 0, //is_free is not queried
+                })?;
+
+                let mut new_start_units = Vec::new();
+                for row in rows {
+                    let unit = row?;
+                    if unit.unit == earlier_unit.unit {
+                        break 'go_up;
```

Comment:
you can use continue key word instead of using labels since only one loop here, but it's Ok like this